ch390/src/esp_eth_mac_ch390.c

```diff
@@ -0,0 +1,907 @@
+        ESP_GOTO_ON_ERROR(ch390_register_read(emac, CH390_NCR, &ncr), err, TAG, "read NCR failed");
+        if (!(ncr & NCR_RST)) {
+            break;
+        }
+        vTaskDelay(pdMS_TO_TICKS(10));
+    }
+    ESP_GOTO_ON_FALSE(to < emac->sw_reset_timeout_ms / 10, ESP_ERR_TIMEOUT, err, TAG, "reset timeout");
+
+    /* power on phy */
+    ESP_GOTO_ON_ERROR(ch390_register_write(emac, CH390_GPR, 0x00), err, TAG, "write GPR failed");
+    /* mac and phy register won't be accesable within at least 1ms */
+    vTaskDelay(pdMS_TO_TICKS(100));
+
+    return ESP_OK;
+err:
+    return ret;
+}
+
+/**
+ * @brief verify ch390 chip ID
+ */
+static esp_err_t ch390_verify_id(emac_ch390_t *emac)
+{
+    esp_err_t ret = ESP_OK;
+    uint8_t id[2];
+    ESP_GOTO_ON_ERROR(ch390_register_read(emac, CH390_VIDL, &id[0]), err, TAG, "read VIDL failed");
+    ESP_GOTO_ON_ERROR(ch390_register_read(emac, CH390_VIDH, &id[1]), err, TAG, "read VIDH failed");
+    ESP_GOTO_ON_FALSE(0x1C == id[1] && 0x00 == id[0], ESP_ERR_INVALID_VERSION, err, TAG, "wrong Vendor ID");
+    ESP_GOTO_ON_ERROR(ch390_register_read(emac, CH390_PIDL, &id[0]), err, TAG, "read PIDL failed");
+    ESP_GOTO_ON_ERROR(ch390_register_read(emac, CH390_PIDH, &id[1]), err, TAG, "read PIDH failed");
+    ESP_GOTO_ON_FALSE(0x91 == id[1] && 0x51 == id[0], ESP_ERR_INVALID_VERSION, err, TAG, "wrong Product ID");
+    return ESP_OK;
+err:
+    return ret;
+}
+
+/**
+ * @brief default setup for ch390 internal registers
+ */
+static esp_err_t ch390_setup_default(emac_ch390_t *emac)
+{
+    esp_err_t ret = ESP_OK;
+    /* disable wakeup */
+    ESP_GOTO_ON_ERROR(ch390_register_write(emac, CH390_NCR, 0x00), err, TAG, "write NCR failed");
+    ESP_GOTO_ON_ERROR(ch390_register_write(emac, CH390_WCR, 0x00), err, TAG, "write WCR failed");
+    /* stop transmitting, enable appending pad, crc for packets */
+    ESP_GOTO_ON_ERROR(ch390_register_write(emac, CH390_TCR, 0x00), err, TAG, "write TCR failed");
+    /* stop receiving, no promiscuous mode, no runt packet(size < 64bytes), receive all multicast packets */
+    /* discard long packet(size > 1522bytes) and crc error packet, enable watchdog */
+    ESP_GOTO_ON_ERROR(ch390_register_write(emac, CH390_RCR, RCR_DIS_CRC | RCR_ALL), err, TAG, "write RCR failed");
+    /* retry late collision packet, at most two transmit command can be issued before transmit complete */
+    ESP_GOTO_ON_ERROR(ch390_register_write(emac, CH390_TCR2, TCR2_RLCP), err, TAG, "write TCR2 failed");
+    /* enable auto transmit */
+    ESP_GOTO_ON_ERROR(ch390_register_write(emac, CH390_ATCR, ATCR_AUTO_TX), err, TAG, "write ATCR failed");
+    /* generate checksum for UDP, TCP and IPv4 packets */
+    ESP_GOTO_ON_ERROR(ch390_register_write(emac, CH390_TCSCR, TCSCR_IPCSE | TCSCR_TCPCSE | TCSCR_UDPCSE), err, TAG, "write TCSCR failed");
+    /* disable check sum for receive packets */
+    ESP_GOTO_ON_ERROR(ch390_register_write(emac, CH390_RCSCSR, 0x00), err, TAG, "write RCSCSR failed");
```

Comment:
Do you know why we cannot enable check sum of receive packets? I always get error status `0x15` of enable it